### DISCUSSION TITLE
merging state from same component that is included twice on a page

### DISCUSSION THREAD

**Discussion by NielsvanHoof** [livewire/livewire#4759] [1. Help] [answered]
Hey, I have a page that includes the same component twice in a page using a loop 

```
<div>
    @for($i = 0; $i < 2; $i++)
        <livewire:inputs.components.textblock :index="$i" wire:key="{{ $i }}" />
    @endfor
</div>
```

this component takes an index that's used inside of the `blade` file to build the following html layout 

```
        <div class="form-control">
                        <label for="textblock_{{$index}}_introTxt" class="label">
                            <span class="label-text">Intro text</span>
                        </label>
                        <textarea wire:model.debounce.500ms='state.{{$index}}.intro' type="text"
                                  name="textblock_{{$index}}_introTxt"
                                  id="textblock_{{$index}}_introTxt"
                                  required class="textarea textarea-bordered">

                        </textarea>
                    </div>
         </div

--- another text area here
```

This gives me two textarea's each with a different `id` e.g. `textblock_0_introTxt` and `textblock_1_intoTxt`

the back end of said component is build like this 

```
<?php

namespace App\Http\Livewire\Inputs\Components;

use Illuminate\Contracts\Foundation\Application;
use Illuminate\Contracts\View\Factory;
use Illuminate\Contracts\View\View;
use Livewire\Component;

class Textblock extends Component
{
    public int $index = 0;

    public array $state = [];

    protected array $rules = [
        'state.*.intro' => ['required', 'string'],
        'state.*.main_text' => ['required', 'string'],
    ];

    protected array $messages = [
        'state.*.intro.required' => 'Please enter a large text',
        'state.*.main_text.required' => 'Please enter a small text',
    ];

    public function render(): Factory|View|Application
    {
        return view('livewire.inputs.components.textblock');
    }
}
```

The most important thing here is the `state` array, this is where each of my textarea's inputs should be added to by id.

the current behavior is as follows: 

I type the following content inside of textarea_0 : "test content 0"

livewire updates the state to the following : 

```
    "state" => array:1 [
      0 => array:1 [
        "intro" => "test"
      ]
    ]
```
This is correct, but when I start typing inside of textarea_1 the following happens to the `state` array:

```
   "state" => array:1 [
      1 => array:1 [
        "intro" => "test 2"
      ]
    ]
```
This time the array is overwriten without appending to it, the expected behavior should be the following: 

```
    "state" => array: [
      0 => array:1 [
        "intro" => "test"
      ],
      1 => array:1 [
        "intro" => "test 2"
      ]
    ]
```

Any idea how to accomplish this in php/livewire?

**Comment by silnarm** [comment]
Ok. So what you are currently doing is creating two components, of the same type, but two separate distinct entities.
Each of those components has its own `$state` array, and they each are filling their own. So there is no overriding happening, just two arrays on two components. The index you are passing in is used for the key, so that looks right, but you are seeing two different arrays in your updated hooks.
I think you can probably much more sensibly do what you want with just one component, but you might need to elaborate a little more on what you are trying to do.

**Comment by NielsvanHoof** [comment]
Hey, it's a POC for our new CMS we're prototyping. it features a sort of "mini" page builder with pre-defined components (coming from a seperate next js app), in this case a "page" can have two distinct textblocks each with unique content. I hope this explains it a bit more

**Comment by joshhanley** [comment]
@NielsvanHoof I wouldn't use Livewire components for this, I'd make blade components for each of the textblocks (or any other content blocks), but then have a Livewire component which wraps the whole content builder, so that way all state is in the one Livewire component.

Livewire wasn't designed to be used a small seperate Livewire components for each input field.

Hope this helps!

**Reply by NielsvanHoof** [reply]
Hey, the content blocks are included inside of a layout (since the layout determines the components one has access to) e.g. a "blog" layout has different components than a "about" layout. 

So for example I have a "landingspage"  livewire component that includes the different inputs (banner, textblock, contactblock e.t.c.) these are at the moment all livewire components. 

From what I understand you want to have the content blocks turned into normal blade components and have the layout act sort of like a wrapper around them? 

This seems like a easy solution to implement, my only other question would be how I would handle the different inputs, at the moment all of the inputs are wire:binded to their backend "state".

I'm not sure how you would get the values from the blade components and have the wrapper set state from that input without using wire:bind?

If you could help me with this, I think I can figure out the rest myself.

Thanks!

**Reply by joshhanley** [reply]
@NielsvanHoof yeah that's correct!

You can still use `wire:model` with blade components, like `<x-my-component wire:model="myProperty" />`

Hope this helps!

**Comment by NielsvanHoof** [comment, marked as answer]
Managed to fix my using using the methodes discussed above, with the layout acting as a master, passing the state to blade components.
